omniscidb/ResultSetRegistry/ResultSetTableToken.cpp

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ResultSetTableToken.h"
+#include "ResultSetRegistry.h"
+
+namespace hdk {
+
+ResultSetTableToken::ResultSetTableToken(TableInfoPtr tinfo,
+                                         size_t row_count,
+                                         std::shared_ptr<ResultSetRegistry> registry)
+    : tinfo_(tinfo), row_count_(row_count), registry_(registry) {
+  CHECK(registry);
+  CHECK_EQ(getSchemaId(tinfo_->db_id), registry->getId());
+}
+
+ResultSetTableToken::~ResultSetTableToken() {
+  reset();
+}
+
+ResultSetPtr ResultSetTableToken::resultSet(size_t idx) const {
+  CHECK_LT(idx, resultSetCount());
+  return registry_->get(*this, idx);
+}
+
+void ResultSetTableToken::reset() {
+  if (!empty()) {
+    registry_->drop(*this);
+    registry_.reset();
```

Comment:
Tokens make a call to the registry in its destructor and that is the reason to hold a `shared_ptr<ResultSetRegistry>`. When a token is reset, it doesn't need to hold the pointer anymore. It doesn't mean we destroy the registry because `DataMgr` and other tokens also keep it alive. I added this `reset` method so that `RelAlgExecutor` could clean up temporary tables it doesn't need anymore (it doesn't do this clean-up yet, but hopefully will do it someday).

Reply:
Oh, I misread it and thought it was a call to some registry method, never mind!